### PR TITLE
HeavyFlavorAnalysis/*Decay redesigned using template interfaces

### PR DESCRIPTION
**Pull request by ronchese**
#### PR description:

The BPH reconstruction package (doc. in https://twiki.cern.ch/twiki/bin/view/CMS/BPHReco_V2 , to be updated with new features)
in HeavyFlavorAnalysis/RecoDecay and HeavyFlavorAnalysis/SpecificDecay has been redesigned using templated interfaces for common decay patterns.

The reconstruction of the decay B+- -> Psi2S K+- with Psi2S -> JPsi pi+ pi- and JPsi -> mu+mu- has been added.

#### PR validation:

Tested on a UL MINIAOD file (Charmonium dataset), same output as previous code.
All test in runtheMatrix passed.

### CODE REVIEW

**Comment by perrotta** [comment]
@ronchese could you please add a subject to the PR title, so that people can realize by reading it what is actually "redesigned" with this PR?

**Comment by mmusich** [comment]
please test

**Comment by ronchese** [comment]
"conflicting" PR 38525 closed

**Comment by ronchese** [comment]
Hi Andrea & Marco
this one does not require 38540, and makes sense also standalone (it's an improvement, anyway), on the other side 38540 reuires some of the features implemented here.

File: HeavyFlavorAnalysis/SpecificDecay/src/BPHDecayConstrainedBuilderBase.cc [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  return;
+}
+
+void BPHDecayConstrainedBuilderBase::setConstr(bool flag) {
+  if ((flag == massConstr) && !mfSelForce)
+    return;
+  outdated = true;
+  massConstr = flag;
+  delete mFitSel;
+  mFitSel = massFitSelector(mFitSel->getMassMin(), mFitSel->getMassMax());
```
Comment (perrotta) [inline HeavyFlavorAnalysis/SpecificDecay/src/BPHDecayConstrainedBuilderBase.cc:92]: `mFitSel` used after deletion (see also static analyzer report)

File: HeavyFlavorAnalysis/RecoDecay/src/BPHDecayVertex.cc [hunk 2]
```diff
@@ -122,34 +126,42 @@ const reco::Vertex& BPHDecayVertex::vertex(VertexFitter<5>* fitter,
...
   map<const reco::Candidate*, reco::TransientTrack*>::const_iterator iend = ttMap.end();
   return (iter != iend ? iter->second : nullptr);
 }
 
 /// retrieve EventSetup
-const edm::EventSetup* BPHDecayVertex::getEventSetup() const { return evSetup; }
+const BPHEventSetupWrapper* BPHDecayVertex::getEventSetup() const { return evSetup; }
 
 const string& BPHDecayVertex::getTrackSearchList(const reco::Candidate* cand) const {
   static string dum = "";
```
Comment (perrotta) [inline HeavyFlavorAnalysis/RecoDecay/src/BPHDecayVertex.cc:167]: No reason to make it `static` (as also pointed out by the static analyzer)
Why not simply `return "";` at L 171?

File: HeavyFlavorAnalysis/SpecificDecay/interface/BPHX3872ToJPsiPiPiBuilder.h [hunk 3]
```diff
@@ -12,102 +12,57 @@
...
-  BPHMassSelect* massSel;
-  BPHChi2Select* chi2Sel;
-  BPHMassFitSelect* mFitSel;
-
-  bool massConstr;
-  float minPDiff;
-  bool updated;
-
-  std::vector<BPHRecoConstCandPtr> x3872List;
+  ~BPHX3872ToJPsiPiPiBuilder() override {}
```
Comment (perrotta) [inline HeavyFlavorAnalysis/SpecificDecay/interface/BPHX3872ToJPsiPiPiBuilder.h:65]:
```suggestion
  ~BPHX3872ToJPsiPiPiBuilder() override = default;
```

File: HeavyFlavorAnalysis/SpecificDecay/interface/BPHPsi2SToJPsiPiPiBuilder.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    setMassFitRange(3.50, 3.80);
+  }
+
+  // deleted copy constructor and assignment operator
+  BPHPsi2SToJPsiPiPiBuilder(const BPHPsi2SToJPsiPiPiBuilder& x) = delete;
+  BPHPsi2SToJPsiPiPiBuilder& operator=(const BPHPsi2SToJPsiPiPiBuilder& x) = delete;
+
+  /** Destructor
+   */
+  ~BPHPsi2SToJPsiPiPiBuilder() override {}
```
Comment (perrotta) [inline HeavyFlavorAnalysis/SpecificDecay/interface/BPHPsi2SToJPsiPiPiBuilder.h:65]:
```suggestion
  ~BPHPsi2SToJPsiPiPiBuilder() override = default;
```

File: HeavyFlavorAnalysis/SpecificDecay/src/BPHDecayToResTrkTrkSameMassBuilderBase.cc [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      nCollection(negCollection),
+      ptMin(0.0),
+      etaMax(100.0) {}
+
+//--------------
+// Operations --
+//--------------
+void BPHDecayToResTrkTrkSameMassBuilderBase::fillTrkTrkList() {
+  double mTotMax = massSel->getMassMax();
+  double mTpnMax = mTotMax - (0.8 * rMass);
```
Comment (perrotta) [inline HeavyFlavorAnalysis/SpecificDecay/src/BPHDecayToResTrkTrkSameMassBuilderBase.cc:78]: This can be moved inside the `if` at L85

**Review by perrotta — COMMENTED** [review]
Very limited review, based on the static analyzer report, and on a quick visual scan of the code

File: HeavyFlavorAnalysis/RecoDecay/src/BPHDecayVertex.cc (hunk 2 above)
Comment (ronchese) [reply HeavyFlavorAnalysis/RecoDecay/src/BPHDecayVertex.cc:167]: The functions return by reference, if not making that static or doing a
`return ""`
the compiler complains about "returning reference to temporary"

Comment (perrotta) [reply HeavyFlavorAnalysis/RecoDecay/src/BPHDecayVertex.cc:167]: Correct...
What about `static const` ?

File: HeavyFlavorAnalysis/SpecificDecay/src/BPHDecayToResTrkTrkSameMassBuilderBase.cc (hunk 5 above)
Comment (ronchese) [reply HeavyFlavorAnalysis/SpecificDecay/src/BPHDecayToResTrkTrkSameMassBuilderBase.cc]: done

**Comment by mmusich** [comment]
please test

**Comment by mmusich** [comment]
@cms-sw/orp-l2 are you happy with this PR now?
It would be good to integrate it in order to have https://github.com/cms-sw/cmssw/pull/38540 integrated as well.

**Pull request merged**